include/audio_backend/__asio_audio_device.h

```diff
@@ -0,0 +1,591 @@
+    {
+    case ASIOSTInt32LSB: _write = [this](long index) { write<int32_t>(index); }; break;
+    case ASIOSTInt24LSB: _write = [this](long index) { write<packed24_t>(index); }; break;
+    case ASIOSTInt16LSB: _write = [this](long index) { write<int16_t>(index); }; break;
+    case ASIOSTFloat32LSB: _write = [this](long index) { write<float>(index); }; break;
+    case ASIOSTFloat64LSB: _write = [this](long index) { write<double>(index); }; break;
+
+    default: throw audio_device_exception("ASIO native sample type not supported: " + _sample_type);
+    }
+  }
+
+  template <typename _SampleType>
+  void write (long index) {
+
+    auto& out = *_io.output_buffer;
+    for (int channel = 0; channel < _num_outputs; ++channel) {
+      const auto buffer = static_cast<__asio_sample<_SampleType>*>(_asio_buffers[_num_inputs + channel].buffers[index]);
+      for (int frame = 0; frame < out.size_frames(); ++frame) {
+        buffer[frame] = {out(frame, channel)};
+      }
+    }
+  }
+
+  static void sample_rate_changed(ASIOSampleRate) {}
+
+  static long asio_message(long selector, long, void*, double*) {
+    switch (selector)
+    {
+    case kAsioSupportsTimeInfo: return ASIOTrue;
+    case kAsioEngineVersion: return 2;
+    case kAsioSelectorSupported: return ASIOFalse;
+    default: return ASIOFalse;
+    }
+  }
+
+  static void buffer_switch(long index, ASIOBool) {
+    instance()->on_buffer_switch(index);
+  }
+
+  void on_buffer_switch(long index) {
+
+    _read(index);
+    _user_callback(*this, _io);
+    _write(index);
+
+    _asio->outputReady();
+  }
+
+  static ASIOTime* buffer_switch_time_info (ASIOTime* time, long index, ASIOBool) {
+    return instance()->on_buffer_switch(time, index);
+  }
+
+  ASIOTime* on_buffer_switch(ASIOTime* time, long index) {
+
+    _read(index);
+    _user_callback(*this, _io);
+    _write(index);
+
+    time->timeInfo.flags = kSystemTimeValid | kSamplePositionValid;
+    time->timeInfo.samplePosition = _sample_position;
+    _sample_position += _buffer_size;
+    time->timeInfo.systemTime = timeGetTime() * 1'000'000;
+    return time;
+  }
+
+  CComPtr<IASIO> _asio;
+  const string _name;
+  device_id_t _id;
+  long _num_inputs = 0;
+  long _num_outputs = 0;
+  vector<sample_rate_t> _sample_rates;
+  vector<buffer_size_t> _buffer_sizes;
+  buffer_size_t _buffer_size = 0;
+  bool _running = false;
+
+  ASIOSampleType _sample_type;
+  ASIOSamples _sample_position = 0;
+
+  using __asio_callback_t = function<void(audio_device&, audio_device_io<float>&)>;
+  __asio_callback_t _user_callback;
+  audio_device_io<__asio_common_sample_type> _io;
+  vector<float> _input_samples;
+  vector<float> _output_samples;
+
+  using fill_buffers = function<void(long index)>;
+  fill_buffers _read;
+  fill_buffers _write;
+
+  vector<ASIOBufferInfo> _asio_buffers;
+  ASIOCallbacks _asio_callbacks;
+
+  static audio_device* instance(audio_device* d = nullptr) {
+    static audio_device* device = d;
```

Comment:
Does this mean that you can never communicate with more than one asio device in the lifetime of a single process?  What if I want to shut ASIO down and reinitialize it?  What if my device gets destructed?  This pointer will become invalid.  Can we come up with a way to do this that does not store an instance pointer into static memory?

Reply:
I'm open to suggestions :)
The problem is that this is a header-only implementation & the ASIO instance needs to be stored somewhere. ODR is violated then multiple translation units inlude <audio>
I'm sure we can do better than the current clunky implementation, I'm just not sure how without having to link as a lib. Could check the value passed in every time & just overwrite if non-null

Reply:
> Can we come up with a way to do this that does not store an instance pointer into static memory?

Could be tricky. ASIO callbacks don't have a back-pointer to the caller.

Reply:
I've addressed some of these concerns in 9056a6f
Different ASIO devices can be used in the lifetime of the process. However, it is still not possible to use more than one ASIO driver concurrently.